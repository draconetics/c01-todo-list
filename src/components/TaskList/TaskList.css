.tasklist{

}

.tasklist__form{
    background-color: #fcf9f9;
    padding:2rem;
    text-align: center;
}

.tasklist__form input:focus{
    outline: none;
    border:1px solid var(--primary-color);
}

.tasklist__content ul{
    display:flex;
    flex-wrap: wrap;
    justify-content: center;
}

.tasklist__content ul li{
    padding: 1rem;
    border-bottom: 1px solid rgba(var(--primary-color),1);
}
.tasklist__content ul:hover{
    background-color: rgba(var(--primary-color),1);
    color:var(--white);
}

.tasklist__content ul li:nth-child(2){
    flex:1;
    text-align: center;
}

.modal{
    width: 100vw;
    height: 100vh;
    position: absolute;
    top:0;
    left: 0;
    background-color: rgba(var(--primary-color),0.5);
    z-index:100;
    opacity:0;
    pointer-events: none;
    transition: opacity .5s ease;
}
.modal.show{
    opacity: 1;
    pointer-events: all;
}

.modal .modal__content{
    max-width: 400px;
    position:absolute;
    top:50%;
    left:50%;
    transform: translate(-50%,-50%);
    background-color: var(--white);
    border: 1px solid rgba(var(--primary-color),1);
    border-radius: 2px;
    box-shadow: 1px 1px 4px 1px rgba(0,0,0,.2);
}

.modal .modal__title{
    background-color: rgba(var(--primary-color),1);
    padding: 1rem;
    color:var(--white);
}

.modal .modal__form{
    padding: 1rem;
}
