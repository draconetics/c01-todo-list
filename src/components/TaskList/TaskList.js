import React from 'react';
import PropTypes from 'prop-types';
import './TaskList.css'

class TaskList extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      task: '',
      editTask:{
        _id:'',
        task:''
      }
    };
    this.modalRef=React.createRef();
  }

  componentDidMount() {
    const {fetchTaskList} = this.props;
    fetchTaskList()
  }

  saveTask(e) {
    e.preventDefault();
    console.log("save")
    const {saveTask} = this.props;
    const {task} = this.state;
    saveTask(task);
    this.setState({task:''})
  }

  updateTask(e){
    e.preventDefault();
    const {updateTask} = this.props;
    const {editTask} = this.state;
    updateTask(editTask);
    this.modalRef.current.classList.toggle('show');
  }

  handleInput(e){
    this.setState({task: e.target.value})
    
  }

  handleEditTask(e){
    e.preventDefault();
    const {editTask} = this.state;
    this.setState({editTask:{...editTask,task:e.target.value}})

  }

  editTask(item){
    console.log(item)
    this.modalRef.current.classList.toggle('show');
    this.setState({editTask:item})
    
  }

  cancel(e){
    e.preventDefault();
    console.log(this.modalRef.current)
    this.modalRef.current.classList.toggle('show');
  }

  deleteTask(id){
    const { deleteTask } = this.props;
    deleteTask(id);
  }

  render = () => {
    const { task, editTask } = this.state;
    const { taskList } = this.props;

    return (
      <div className="tasklist container" data-test="TaskList">
        <form className="tasklist__form" onSubmit={(e)=>this.saveTask(e)}>
          <input type="text"  onChange={e=>this.handleInput(e)} value={task}/>
          <button type="submit" className="btn">Add new task</button>
        </form>
        <div className="tasklist__content">
          {taskList &&
            taskList.map((item, index) => {
              return (
                <ul key={index}>
                  <li>{index+1}</li>
                  <li>{item.task}</li>
                  <li><button className="btn" onClick={()=>this.editTask(item)}>Edit</button></li>
                  <li><button className="btn" onClick={()=>this.deleteTask(item._id)}>Delete</button></li>
                </ul>
              );
            })}
        </div>
        <div className="modal" ref={element=>this.modalRef.current = element}>
            <div className="modal__content">
              <div className="modal__title">Edit your item</div>
              <form className="modal__form" onSubmit={(e)=>this.updateTask(e)}>
                <input type="text" value={editTask.task} onChange={(e)=>this.handleEditTask(e)} />
                <button type="submit" className="btn">save</button>
                <button className="btn" onClick={(e)=>this.cancel(e)}>cancel</button>
              </form>
            </div>
        </div>
      </div>
    );
  }; // end render
}

const item = {
  _id: PropTypes.any,
  task: PropTypes.string,
};
TaskList.propTypes = {
  taskList: PropTypes.arrayOf(PropTypes.shape(item)).isRequired,
  fetchTaskList: PropTypes.func.isRequired,
  updateTask: PropTypes.func.isRequired,
  saveTask: PropTypes.func.isRequired,
  deleteTask: PropTypes.func.isRequired,
};
export default TaskList;
