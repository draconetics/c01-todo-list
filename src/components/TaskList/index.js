import TaskList from './TaskList'
import { saveTask, updateTask, fetchTaskList, deleteTask } from '../../store/actions'
import { connect } from 'react-redux';


  const mapStateToProps = state =>{
      return {
          taskList: state.taskReducer.taskList
      }
  }
 
  const mapDispatchToProps = dispatch => {
    return {
        fetchTaskList: ()=>dispatch(fetchTaskList),
        updateTask: (item)=>dispatch(updateTask(item)),
        saveTask: (task) =>dispatch(saveTask(task)),
        deleteTask: (id) => dispatch(deleteTask(id)),
    };
  }; 
  
  export default connect(mapStateToProps, mapDispatchToProps)(TaskList);