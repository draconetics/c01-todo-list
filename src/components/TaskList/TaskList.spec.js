import React from 'react';
import { mount, shallow } from 'enzyme';
import TaskList from './TaskList'

describe("#TaskList",()=>{
    describe("#TaskList Render", ()=>{
        
        let wrapper;

        beforeAll((done)=>{
            let props = {
                taskList:[{_id:1,task:"example01"},{_id:2,task:"example02"}], 
                fetchTaskList: jest.fn(),
                updateTask: jest.fn(),
                saveTask:jest.fn(), 
                deleteTask:jest.fn()
            }

            wrapper = mount(<TaskList {...props}/>)
            done();
        })

        it ('should render component',()=>{
            const taskList =  wrapper.find('[data-test="TaskList"]');
            expect(taskList).toHaveLength(1);
            expect(wrapper).toMatchSnapshot();
        });
    })
})