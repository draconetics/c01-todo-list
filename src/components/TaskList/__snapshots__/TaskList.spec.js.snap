// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#TaskList #TaskList Render should render component 1`] = `ReactWrapper {}`;
